### ISSUE TITLE
Warning when templated conversion function converts to different template

### ISSUE THREAD

**Issue by BMBurstein** [llvm/llvm-project#79434]
Minimal sample:

```c++
#include <type_traits>

template <bool B>
struct S {
    // #1
    template <bool C = B, class = std::enable_if_t<!C>>
    operator S<true>() const { return {}; }

    // #2
    // operator S<true>() const { return {}; }
};

int main() {
    // S<true> b1;
    // S<false> b2;
}
```

Even though the conversion function is only enabled to convert from `S<false>` to `S<true>`, it still produces a warning:

> warning: conversion function converting 'S<true>' to itself will never be used [-Wclass-conversion]

Note that enabling #2 instead of #1, which seems more likely to trigger this warning, in fact does not trigger it.
